### PR TITLE
Add promote task

### PR DESCRIPTION
**Pull request by ryuyu**
Also renames the code push task. Fixes vsts task lib reference

@lostintangent @geof90 @shishirx34

### CODE REVIEW

**Comment by geof90** [comment]
LGTM for now! Some future work that could be valuable

1. Rename `code-push-vsts-task` folder to `codepush-release` for consistency

2. Have `buildCommand`, `executeCommandAndHandleResult` etc helper functions be shared

3. Have `login` and `logout` code be shared between the two tasks

4. Have the test code for `login` and `logout` be shared.

File: Tasks/codepush-promote/task.json [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            "type": "string",
+            "label": "App Name",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "Name of the app you want to release an update for"
+        },
+        {
+            "name": "sourceDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Should we label this as "Source Deployment Name"?

File: Tasks/codepush-promote/task.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            "defaultValue": "",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Let's make "Staging" the default value.

File: Tasks/codepush-promote/task.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            "required": false,
+            "helpMarkDown": "Name of the deployment under the app which to promote from",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Let's make this "Name of the deployment you want to promote the latest release from"

File: Tasks/codepush-promote/task.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                "Production": "Production"
+            },
+            "properties": {
+                "EditableOptions": "True"
+            }
+        },
+        {
+            "name": "targetDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Let's label this as "Target Deployment Name"

File: Tasks/codepush-promote/task.json [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            "defaultValue": "",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Let's make "Production" the default value

File: Tasks/codepush-promote/task.json [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            "helpMarkDown": "Access key used to authenticate with the CodePush service",
+            "visibleRule": "authType = AccessKey"
+        },
+        {
+            "name": "appName",
+            "type": "string",
+            "label": "App Name",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "Name of the app you want to release an update for"
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Should this be "Name of the app you want to promote a release for"

File: Tasks/codepush-promote/task.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "name": "sourceDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
+            "defaultValue": "",
+            "required": false,
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: This should be required

File: Tasks/codepush-promote/task.json [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                "EditableOptions": "True"
+            }
+        },
+        {
+            "name": "targetDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Name of the deployment under the app which you want to push the release to",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: How about "Name of the deployment you want to promote the release to"

File: Tasks/codepush-promote/package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "vsts-task-code-push-promote",
+  "version": "1.0.0",
+  "description": "A VSTS deploy task for releasing app updates to the CodePush service",
```
Comment (lostintangent) [inline Tasks/codepush-promote/package.json]: Copy the description from the task.json

File: Tasks/codepush-promote/codepush-promote.ps1 [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    [string]$targetDeploymentName
+) 
+  
+$env:INPUT_accessKey = $accessKey
+$env:INPUT_serviceEndpoint = $serviceEndpoint
+$env:INPUT_appName = $appName
+$env:INPUT_sourceDeploymentName = $sourceDeploymentName
+$env:INPUT_targetDeploymentName = $targetDeploymentName
+
+node CodePush.js
```
Comment (lostintangent) [inline Tasks/codepush-promote/codepush-promote.ps1]: Shouldn't this be "nose codepush-promote.js"?

File: Tasks/code-push-vsts-task/CodePush.js [hunk 11]
```diff
@@ -1,5 +1,5 @@
 var path = require("path");
-var tl = require("vso-task-lib");
```
Comment (lostintangent) [inline Tasks/code-push-vsts-task/CodePush.js]: Should we rename this directory while we're at it?

File: Tasks/code-push-vsts-task/task.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
     "id": "b69d19a0-7f6d-11e5-a86a-c199aa7a6cf0",
-    "name": "CodePush",
-    "friendlyName": "CodePush",
+    "name": "codepush-release",
+    "friendlyName": "CodePush - Release",
     "description": "A VSTS build task for releasing app updates to the CodePush service",
```
Comment (lostintangent) [inline Tasks/code-push-vsts-task/task.json]: Can we change "build" to "build/release"? Now that we're focusing on release management as well, we shouldn't tie the task to just build.

File: Tasks/codepush-promote/codepush-promote.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    // If function arguments are provided (e.g. during test), use those, else, get user inputs provided by VSTS.
+    var authType = tl.getInput("authType", false);
+    if (authType === "AccessKey") {
+        accessKey = tl.getInput("accessKey", true);
+    } else if (authType === "ServiceEndpoint") {
+        var serviceAccount = tl.getEndpointAuthorization(tl.getInput("serviceEndpoint", true));
+        accessKey = serviceAccount.parameters.password;
+    }
+
+    accessKey = accessKey;
```
Comment (lostintangent) [inline Tasks/codepush-promote/codepush-promote.js]: Why is this necessary?

File: Tasks/codepush-promote/codepush-promote.ps1 [hunk 14]
```diff
@@ -0,0 +1,15 @@
+param (
+    [string]$accessKey,
```
Comment (lostintangent) [inline Tasks/codepush-promote/codepush-promote.ps1]: Don't we need to take in the auth type input as well?

File: Tasks/codepush-promote/task.json [hunk 15]
```diff
@@ -0,0 +1,100 @@
+{
+    "id": "4e901d0e-189b-4403-a030-1d06ff8a3b28",
+    "name": "codepush-promote",
+    "friendlyName": "CodePush - Promote",
+    "description": "A VSTS build task for promoting CodePush releases from one deployment to another",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Change "build" to "build/release".

File: Tasks/codepush-promote/task.json [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        {
+            "name": "targetDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Name of the deployment under the app which you want to push the release to",
+            "options": {
+                "Staging": "Staging",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Sort these

File: Tasks/codepush-promote/task.json [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            "name": "sourceDeploymentName",
+            "type": "pickList",
+            "label": "Deployment Name",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Name of the deployment under the app which to promote from",
+            "options": {
+                "Staging": "Staging",
+                "Production": "Production"
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Sort

File: Tasks/codepush-promote/task.json [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+                "AccessKey": "Access Key"
+            }
+        },
+        {
+            "name": "serviceEndpoint",
+            "type": "connectedService:codepush-auth-key",
+            "label": "Service Endpoint",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "The VSTS Service Endpoint that specifies the identity for Code Push.",
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: How about "A CodePush service endpoint that is configured with the necessary credentials"?

File: Tasks/codepush-promote/task.json [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    "inputs": [
+        {
+            "name": "authType",
+            "type": "pickList",
+            "label": "Authentication Method",
+            "defaultValue": "ServiceEndpoint",
+            "helpMarkDown": "",
+            "options": {
+                "ServiceEndpoint": "Service Endpoint",
+                "AccessKey": "Access Key"
```
Comment (lostintangent) [inline Tasks/codepush-promote/task.json]: Sort these

File: Tasks/codepush-promote/package.json [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+  "main": "CodePush.js",
```
Comment (lostintangent) [inline Tasks/codepush-promote/package.json]: It probably doesn't really matter, but this should be "codepush-promote.js".

**Comment by lostintangent** [comment]
Do you want to take a stab at updating the readme? Maybe add a subheading to the task reference section that defines the options for the release and promote tasks seperately?

File: Tasks/code-push-vsts-task/CodePush.js (hunk 11 above)
Comment (ryuyu) [inline Tasks/code-push-vsts-task/CodePush.js]: I'm not sure what directory you mean here?

File: Tasks/codepush-promote/codepush-promote.js (hunk 13 above)
Comment (ryuyu) [inline Tasks/codepush-promote/codepush-promote.js]: it isn't. Leftover from previous version. Since Removed.

File: Tasks/codepush-promote/codepush-promote.ps1 (hunk 14 above)
Comment (ryuyu) [inline Tasks/codepush-promote/codepush-promote.ps1]: Good catch.

File: README.md [hunk 21]
```diff
@@ -34,24 +34,52 @@ This task can be used with either VSTS or TFS 2015 on-prem servers (see below) a
...
+
+The CodePush Promote task includes the following options which can be used to customize your release:
+
+1. **Access Key** (String, required) or **Service Endpoint** - The access key to use to authenticate with the CodePush service. This value can be generated using the [CodePush CLI](https://github.com/Microsoft/code-push/tree/master/cli#authentication) and stored away in vsts using the service endpoint included with the extension.
+
+2. **App Name** (String, Required) - The name of the app that has the deployments you are targetting for promotion.
+
+3. **Source Deployment Name** (String) - Name of the deployment you want to promote the latest release from. Defaults to `Staging`.
+
+4. **Destination Deployment Name** (String) - Name of the deployment you want to promote the release to. Defaults to `Staging`.
```
Comment (lostintangent) [inline README.md]: The default for this would be "Production" right?

File: Tasks/code-push-vsts-task/CodePush.js (hunk 11 above)
Comment (lostintangent) [inline Tasks/code-push-vsts-task/CodePush.js]: Doh! Sorry I clicked the wrong line :) I was referring to the directory that includes the release task. It might make sense to rename it to "codepush-release" for cleanliness sake.

File: Tasks/codepush-promote/codepush-promote.js [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    } else if (authType === "ServiceEndpoint") {
+        var serviceAccount = tl.getEndpointAuthorization(tl.getInput("serviceEndpoint", true));
+        accessKey = serviceAccount.parameters.password;
+    }
+
+    appName = appName || tl.getInput("appName", true);
+    sourceDeploymentName = sourceDeploymentName || tl.getInput("sourceDeploymentName", true);
+    targetDeploymentName = targetDeploymentName || tl.getInput("targetDeploymentName", true);
+
+    if (!accessKey) {
```
Comment (lostintangent) [inline Tasks/codepush-promote/codepush-promote.js]: In what scenario could the access key be null here? Don't both the service endpoint and task require the access key to be populated?

File: README.md [hunk 23]
```diff
@@ -34,24 +34,52 @@ This task can be used with either VSTS or TFS 2015 on-prem servers (see below) a
 
 8. Run your CodePush-ified app to see the change that was automatically deployed!
 
+### Saving Your Access Key
```
Comment (lostintangent) [inline README.md]: Maybe call this "Configuring Your CodePush Credentials"?

File: README.md [hunk 24]
```diff
@@ -34,24 +34,52 @@ This task can be used with either VSTS or TFS 2015 on-prem servers (see below) a
...
+
+1. Generate your access key as described above
+
+2. Go into your Visual Studio Team Services or TFS project and click on the gear in the upper right hand corner
```
Comment (lostintangent) [inline README.md]: "...gear icon in the upper right corner"

File: README.md [hunk 25]
```diff
@@ -34,24 +34,52 @@ This task can be used with either VSTS or TFS 2015 on-prem servers (see below) a
...
+
+3. Click on the **Services** tab
+
+4. Click on **New Service Endpoint** and select **CodePush**
+
+5. Give the new endpoint a name and enter the access key we generated earlier.
```
Comment (lostintangent) [inline README.md]: "...access key you generated earlier"

File: README.md (hunk 21 above)
Comment (ryuyu) [inline README.md]: Yes. Fixed.

File: Tasks/codepush-promote/codepush-promote.js (hunk 22 above)
Comment (ryuyu) [inline Tasks/codepush-promote/codepush-promote.js]: This could possible be null if this module were used as a module, such as in the test (ie authType is not defined). We'll just check for it here so we can short circuit the command call if we don't have one.

**Comment by lostintangent** [comment]
Throughout the readme, we need to change all references to the "CodePush" task to "CodePush - Release", and also change the mentions of the extension including a single task to the plural form as appropriate.
